src/formats/XYZ.cpp

```diff
@@ -252,6 +304,115 @@ std::string write_extended_comment_line(const Frame& frame) {
+
+    const auto& first_atom = frame[0];
+    for (const auto& property: first_atom.properties()) {
+        if (!is_valid_property_name(property.first)) {
+            warning(
+                "Extended XYZ", "'{}' is not a valid property name for extended "
+                "XYZ, is will not be saved",
+                property.first, 0
+            );
+            partially_defined_already_warned.insert(property.first);
+            continue;
+        }
+
+        if (property.second.kind() == Property::STRING) {
+            if (should_be_quoted(property.second.as_string())) {
+                warning(
+                    "Extended XYZ", "string value for property '{}' on atom {} "
+                    "can not be be saved as an atomic property",
+                    property.first, 0
+                );
+                continue;
+            }
+        }
+
+        all_properties.emplace(property.first, property.second.kind());
+    }
+
+    for (size_t i=1; i<frame.size(); i++) {
+        const auto& atom = frame[i];
+
+        // warn for properties defined on this atom but not on others
+        for (const auto& property: atom.properties()) {
+            if (all_properties.count(property.first) == 0) {
+                if (partially_defined_already_warned.count(property.first) == 0) {
+                    warning(
+                        "Extended XYZ",
+                        "property '{}' is only defined for a subset of atoms, it will not be saved",
+                        property.first
+                    );
+                    partially_defined_already_warned.insert(property.first);
+                }
+            }
+        }
+
+        auto to_remove = std::vector<std::string>();
+        for (const auto& property: all_properties) {
+            auto current_property = atom.get(property.first);
+            if (!current_property) {
+                // this property was present for all atoms until now, but
+                // not on the current one
+                warning(
+                    "Extended XYZ",
+                    "property '{}' is only defined for a subset of atoms, it will not be saved",
+                    property.first
+                );
+
+                to_remove.push_back(property.first);
+                continue;
+            }
+
+            if (current_property.value().kind() != property.second) {
+                warning(
+                    "Extended XYZ",
+                    "property '{}' is defined with different types on different atoms, if will not be saved",
+                    property.first
+                );
+                partially_defined_already_warned.insert(property.first);
+                to_remove.push_back(property.first);
+                continue;
+            }
+        }
+
+        for (const auto& name: to_remove) {
+            all_properties.erase(name);
+        }
+    }
+
+    auto results = properties_list_t();
```

Comment:
Would this make things better?
```suggestion
    auto results = properties_list_t();
    results.reserve(all_properties.size());
```